packages/core/src/index.ts

```diff
@@ -122,73 +128,84 @@ export default class Engine<Name extends string = string> {
 				this.publicParsedRules[name] = rule as RuleNode<Name>
 			}
 		}
+
+		this.publicSituation = {} as Situation<Name>
 	}
 
 	resetCache() {
 		this.cache = emptyCache()
 	}
 
 	setSituation(
-		situation: Partial<Record<Name, PublicodesExpression | ASTNode>> = {},
-		options: { keepPreviousSituation?: boolean } = {},
+		situation: Situation<Name> = {},
+		options: {
+			keepPreviousSituation?: boolean
+			safeMode?: boolean
+		} = {},
 	) {
 		this.resetCache()
 
 		const keepPreviousSituation = options.keepPreviousSituation ?? false
-
-		Object.keys(situation).forEach((name) => {
-			if (!(name in this.baseContext.parsedRules)) {
-				throw new PublicodesError(
-					'EvaluationError',
-					`Erreur lors de la mise à jour de la situation : ${name} n'existe pas dans la base de règle.`,
-					{ dottedName: name },
-				)
-			}
-			if (this.baseContext.parsedRules[name].private) {
-				throw new PublicodesError(
-					'EvaluationError',
-					`Erreur lors de la mise à jour de la situation : ${name} est une règle privée (il n'est pas possible de modifier une règle privée).`,
-					{ dottedName: name },
-				)
-			}
-		})
-
-		// The situation is implemented as a special sub namespace `$SITUATION`,
-		// present on each non-private rules
-		const situationToParse = Object.fromEntries(
-			Object.entries(situation).map(([nom, value]) => [
-				`[privé] ${nom} . $SITUATION`,
-				value && typeof value === 'object' && 'nodeKind' in value ?
-					{ valeur: value }
-				:	value,
-			]),
+		const safeMode = options.safeMode ?? this.baseContext.safeMode ?? false
+		const previousContext = this.context
+
+		this.context =
+			keepPreviousSituation ? this.context : copyContext(this.baseContext)
+
+		let situationRules = Object.entries(situation).filter(
+			([dottedName, value]) => {
+				const error = this.checkSituationRule(dottedName, value)
+				if (!error) return true
+				if (safeMode) {
+					this.baseContext.logger.error(error.message)
+					return false
+				} else {
+					this.context = previousContext
```

Comment:
question: on est d'accord que dans le cas où `keepPreviousSituation` est `true`, cette assignation ne sert à rien vu que `previousContext === this.context` ?

Reply:
Bien vu. Cette ligne a été ajoutée lorsque `checkSituationRule` faisait également le parsing de la règle et modifiait this.context par effet de bord.